sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDbStateStoreProvider.scala

```diff
@@ -0,0 +1,652 @@
+            FileUtility.createTarFile(dbPath, snapShotFileName)
+            val targetFile = snapshotFile(baseDir, lastVersion)
+            uploadFile(fm, new Path(snapShotFileName), targetFile, sparkConf)
+          }
+          logInfo(
+            s"Creating snapshot file for" +
+              s" ${stateStoreId_.partitionId} took $t1 ms.")
+        } catch {
+          case e: Exception =>
+            logError(s"Exception while creating snapshot $e} ")
+            throw e
+        } finally {
+          f.delete() // delete the tarball
+        }
+      }
+    }
+  }
+
+  /*
+   * Clean up old snapshots and delta files that are not needed any more. It ensures that last
+   * few versions of the store can be recovered from the files, so re-executed RDD operations
+   * can re-apply updates on the past versions of the store.
+   */
+  private[state] def cleanup(files: Seq[WALUtils.StoreFile]): Unit = {
+    try {
+      if (files.nonEmpty) {
+        val earliestVersionToRetain = files.last.version - storeConf.minVersionsToRetain
+        if (earliestVersionToRetain > 0) {
+          val earliestFileToRetain = filesForVersion(files, earliestVersionToRetain).head
+          val filesToDelete = files.filter(_.version < earliestFileToRetain.version)
+          val (_, e2) = Utils.timeTakenMs {
+            filesToDelete.foreach { f =>
+              fm.delete(f.path)
+              val file = new File(rocksDbPath, f.version.toString.toUpperCase(Locale.ROOT))
+              if (file.exists()) {
+                file.delete()
+              }
+            }
+          }
+          logDebug(s"deleting files took $e2 ms.")
+          logInfo(
+            s"Deleted files older than ${earliestFileToRetain.version} for $this: " +
+              filesToDelete.mkString(", "))
+        }
+      }
+    } catch {
+      case NonFatal(e) =>
+        logWarning(s"Error cleaning up files for $this", e)
+    }
+  }
+
+  private def cleanRocksDBBackupInstances(files: Seq[WALUtils.StoreFile]): Unit = {
+    try {
+      if (files.nonEmpty) {
+        val earliestVersionToRetain = files.last.version - storeConf.minVersionsToRetain
+        if (earliestVersionToRetain > 0) {
+          for (v <- (earliestVersionToRetain - 1) to 1) {
+            // Destroy the backup path
+            logDebug((s"Destroying backup version = $v"))
+            RocksDbInstance.destroyDB(getBackupPath(v))
+          }
+        }
+      }
+    } catch {
+      case NonFatal(e) => logWarning(s"Error cleaning up files for $this", e)
+    }
+  }
+
+  // Used only for unit tests
+  private[sql] def latestIterator(): Iterator[UnsafeRowPair] = synchronized {
+    val versionsInFiles = fetchFiles(fm, baseDir).map(_.version).toSet
+    var itr = Iterator.empty
+    if (versionsInFiles.nonEmpty) {
+      val maxVersion = versionsInFiles.max
+      if (maxVersion == 0) {
+        return Iterator.empty
+      }
+      // FIXME assuming maxVersion exists in rocksDB
```

Comment:
Please address this sooner - either fail the query or do workaround. Let's not leave this as undefined behavior.

Reply:
And same here. HDFS state store provider "loads" the version if it's not cached in memory. RocksDB state store provider should guarantee similar one. Let's make it safe.